src/main/java/co/cask/gcp/datastore/source/DatastoreSource.java

```diff
@@ -209,44 +221,73 @@ Schema constructSchema(Entity entity, boolean isIncludeKey, String keyName) {
   }
 
   /**
-   * Since Datastore is schemaless database, returns nullable field schema for the given value type.
-   * For {@link ValueType#NULL} will return default {@link Schema.Type#STRING},
-   * for unsupported types will return null.
+   * Since Datastore is schemaless database, creates field with nullable schema for the given value
+   * based on its value type, for unsupported types returns null.
    *
    * @param name field name
-   * @param value Datastore value type
+   * @param value Datastore value
    * @return CDAP field
    */
   private Schema.Field transformToField(String name, Value<?> value) {
+    Schema schema = createSchema(name, value);
+    return schema == null ? null : Schema.Field.of(name, Schema.nullableOf(schema));
+  }
+
+  /**
+   * Creates CDAP schema based on given Datastore value and its type.
+   * For {@link ValueType#NULL} will return default {@link Schema.Type#STRING},
+   * for unsupported types will return null.
+   *
+   * @param name field name
+   * @param value Datastore value
+   * @return CDAP schema
+   */
+  private Schema createSchema(String name, Value<?> value) {
+    Schema schema = SUPPORTED_SIMPLE_TYPES.get(value.getType());
+
+    if (schema != null) {
+      return schema;
+    }
+
     switch (value.getType()) {
-      case STRING:
-        return Schema.Field.of(name, Schema.nullableOf(Schema.of(Schema.Type.STRING)));
-      case LONG:
-        return Schema.Field.of(name, Schema.nullableOf(Schema.of(Schema.Type.LONG)));
-      case DOUBLE:
-        return Schema.Field.of(name, Schema.nullableOf(Schema.of(Schema.Type.DOUBLE)));
-      case BOOLEAN:
-        return Schema.Field.of(name, Schema.nullableOf(Schema.of(Schema.Type.BOOLEAN)));
-      case TIMESTAMP:
-        return Schema.Field.of(name, Schema.nullableOf(Schema.of(Schema.LogicalType.TIMESTAMP_MICROS)));
-      case BLOB:
-        return Schema.Field.of(name, Schema.nullableOf(Schema.of(Schema.Type.BYTES)));
       case NULL:
         // Datastore stores null values in Null Type unlike regular databases which have nullable type definition
         // set default String type, instead of skipping the field
-        LOG.debug("Unable to determine type, setting default type '{}' for the field '{}'", Schema.Type.STRING, name);
-        return Schema.Field.of(name, Schema.nullableOf(Schema.of(Schema.Type.STRING)));
+        LOG.debug("Field '{}' is null, setting default type '{}'", name, Schema.Type.STRING);
+        return Schema.of(Schema.Type.STRING);
```

Comment:
why not use the Schema.Type.NULL type for this?

Reply:
Agree, changed to NULL.